tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/aux/CherryPick.scala

```diff
@@ -441,22 +448,30 @@ class CherryPick(rewriter: SymbStateRewriter) {
 
       // The awesome property of the oracle is that we do not have to compare the sets directly!
       // Instead, we compare the oracle values.
-      //  (chosen = 1 /\ in(z_i, R) <=> in(c_i, S_1)) \/ (chosen = 2 /\ in(z_i, R) <=> in(d_i, S_2)) \/ (chosen = N <=> elseAssert)
-      def nthIn(elemAndSet: (ArenaCell, ArenaCell), no: Int): TlaEx = {
+      // (chosen = 1 /\ in(z_i, R) <=> in(c_i, S_1)) \/ (chosen = 2 /\ in(z_i, R) <=> in(d_i, S_2)) \/ (chosen = N <=> elseAssert)
+      def nthIn(elemAndSet: (ArenaCell, ArenaCell), no: Int): (TlaEx, TlaEx) = {
         if (elemsOfMemberSets(no).nonEmpty) {
-          tla.ite(tla.apalacheSelectInSet(elemAndSet._1.toNameEx, elemAndSet._2.toNameEx),
+          val inSet = tla.ite(tla.apalacheSelectInSet(elemAndSet._1.toNameEx, elemAndSet._2.toNameEx),
               tla.apalacheStoreInSet(picked.toNameEx, resultCell.toNameEx),
               tla.apalacheStoreNotInSet(picked.toNameEx, resultCell.toNameEx))
+          val unchangedSet = if (rewriter.solverContext.config.smtEncoding == "arrays") {
+            // In the arrays encoding we need to propagate the SSA representation of the array if nothing changes
+            tla.apalacheStoreNotInSet(picked.toNameEx, resultCell.toNameEx)
+          } else {
+            tla.bool(true)
+          }
+          (inSet, unchangedSet)
         } else {
-          tla.apalacheStoreNotInSet(picked.toNameEx, resultCell.toNameEx) // nothing belongs to the set
+          // nothing belongs to the set
+          (tla.apalacheStoreNotInSet(picked.toNameEx, resultCell.toNameEx), tla.bool(true))
         }
       }
 
-      val assertions = toPickFrom.zip(memberSets).zipWithIndex map (nthIn _).tupled
+      val assertions = (toPickFrom.zip(memberSets).zipWithIndex map (nthIn _).tupled).unzip
       // add the cell to the arena
       nextState = nextState.updateArena(_.appendHas(resultCell, picked))
       // (chosen = 1 /\ in(z_i, R) = in(c_i, S_1)) \/ (chosen = 2 /\ in(z_i, R) = in(d_i, S_2))
-      solverAssert(oracle.caseAssertions(nextState, assertions :+ elseAssert))
+      solverAssert(oracle.caseAssertions(nextState, assertions._1 :+ elseAssert, assertions._2 :+ ValEx(TlaBool(true))))
```

Comment:
Extract `assertions._1` and `assertions._2` to `membershipAssertions` and `nonMembershipAssertions`? Just for documentation purposes.

Reply:
Will do, thanks!